### ISSUE TITLE
Broken with recent PIWIK Versions

### ISSUE THREAD

**Issue by phish108** [mgazdzik/plugin-ShortcodeTracker#25]
Are there any plans updating the plugin to more recent versions?

**Comment by mannp** [comment]
I was hoping the same :)

**Comment by mgazdzik** [comment]
Hi,
I've finally found some time to work on plugin and currently I think at least most of things work with Piwik 3.2 . There are probably still few minor things to work on like automated build, fixing tests etc. but hopefully in few days there's gonna be 1.0 release of plugin which will be compatible with latest Piwik. 
Also would be great if you could give a shot to plugin after migration and help verifying if all functionalities are still intact.

cheers!

**Comment by mannp** [comment]
@mgazdzik well hey thats a great surprise and fantastic news. Thanks for taking the time to update it.

For sure once you release it I will configure and give it a test.

**Comment by mgazdzik** [comment]
@mannp I can only apologize for keeping you wait so long, but hope you'll find use for plugin once more with latest Piwik :)
just a minute ago I've released 1.0.0-RC1 version to master, so please feel free of giving it a shot and in case of any problems let me know.

cheers :)

**Comment by mannp** [comment]
@mgazdzik just glad its now done :) thanks again.

So I have cloned the plugin-ShortcodeTracker to ShortcodeTracker in the plugins directory and enabled the plugin, so I presume thats the install method at the moment?

I will need to configure after work, so will try and get it working in the next days and report back.

**Comment by mgazdzik** [comment]
@mannp correct, enabling plugin should do the job for installation

**Comment by mannp** [comment]
Hi @mgazdzik I finally have time to take a look at this after the holidays and had one very basic question.

I have my website and piwik setup through virtual hosts behind a nginx proxy, so do I need to create a new subdomain say redirects.domain.co which uses your vhost config found in the setup section?

Thanks in advance

**Comment by mgazdzik** [comment]
Hi @mannp, glad you have time to check out latest plugin :)
It is not strictly necessary to have separate subdomain, it can be a sub-path in your main site vhost. 
Please check doc available at https://github.com/mgazdzik/plugin-ShortcodeTracker/blob/master/docs/nginx_config.md (especially `rewrite` section).
There is redirection within single domain (presumably Piwik one), but for example you can do something like that

    rewrite "^\/s\/([a-zA-Z0-9]{6})$" https://your-piwik-instance.co/index.php?module=API&method=ShortcodeTracker.performRedirectForShortcode&code=$1;

You put that directive in config for vhost from which you want to handle redirections (for ex. your main page), and regex applied will affect addresses like https://example.com/s/asdQWE to https://your-piwik-instance.co/index.... etc.

So TL;DR - only configuration on vhost that you need to do is for domain which will handle redirections and proxy those requests to proper Piwik API address.

Please let me know if I can elaborate more on this logic.

Cheers,
Michał

**Comment by mannp** [comment]
Hi @mgazdzik I am sorry but I am still confused. I am getting confused with what you mean by 'main site vhost' and 'vhost from which you want to handle redirections (for ex. your **main page**)'

I am expecting to use say **https://shortened.url** which points to some full length link. When the shortened link is clicked it has to be translated by something into the full link, so is that done by my website or is that done by my piwik instance?

> www.domain.org or piwik.domain.org or shorturl.domain.org

Is there an example of what I should be entering into the 'Shortener Url' field in the config? 

I have an nginx proxy feeding my nginx piwik instance through https, so I am expecting the redirect to be added to my nginx vhost config with the existing working config?

Having added the redirect I try and create a shortened code for an external site and I get the following;

> SQLSTATE[42S02]: Base table or view not found: 1146 Table 'piwik.piwik_shortcode' doesn't exist

Again sorry for further questions.

**Comment by mgazdzik** [comment]
@mannp ok, let's clear those things out ;)
In my thinking there are 2 vhosts:
- domain you want to use for redirects (for ex. in my case it's my homepage domain)
- domain on which your piwik is hosted (in my case it's subdomain).

The rule to be pasted into nginx config should affect your redirecting domain vhost.
Example - it should be a full url of that domain, for ex. "https://example.com/short/"

then every shortcode you'll generate will be appended after that URL and returned to you in Piwik interface (for ex. result would be https://example.com/short/abc123")

As for error you've posted it seems that for some reason database tables weren't created when installing plugin. I would recommend removing plugin name occurences from your config.ini.php file (to be precise from "Plugins" and "PluginsInstalled" section). After that when you try to activate your plugin again, it should execute installing methods and create necessary db schema.
Please let me know how did it go, hope it helps :)

cheers!

**Comment by mannp** [comment]
@mgazdzik ok thanks for the extra clarification, its becoming clearer now.

Does this directory on you main domain ("https://example.com/short/") work for external site redirects too?

Lastly, I have piwik and my main website on separate dockers, so I need to share the root of my piwik instance with my website domain to satisfy the 'root /config/www/piwik' of the server {} definition which will reside on my website docker, as the redirecting domain?

**Comment by mgazdzik** [comment]
Hi @mannp, just wanted to reach out and ask if you've maybe managed to kickstart plugin on your Piwik instance?

Cheers,
Michał

**Comment by mannp** [comment]
Hi Michał @mgazdzik I am embarrassed to say I have not had the time with my business and a lot of other things. 

I very much appreciate your time in coding it, so will find the time as soon as I am able.

**Comment by mannp** [comment]
I have tried going back to basics so I am not in conflict with the other webservers I have setup.

I have a nginx docker server with php configured and the following default config;

erver {
        listen 80 default_server;

        listen 443 ssl;

        root /config/www;
        index index.php;

        server_name _;

        rewrite "^\/([a-zA-Z0-9]{6})$"
https://piwik-fqdn/piwik/index.php?module=API&method=ShortcodeTracker.performRedirectForShortcode&code=$1;

        location / {
                index index.php;
        }

        ssl_certificate /config/keys/cert.crt;
        ssl_certificate_key /config/keys/cert.key;

        client_max_body_size 0;

        location ~ \.php$ {
                try_files $uri =404;
                fastcgi_split_path_info ^(.+\.php)(/.+)$;
                fastcgi_pass 127.0.0.1:9000;
                fastcgi_index index.php;
                fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
                include /etc/nginx/fastcgi_params;
        }
}

 With an nginx server acting as a proxy with the following default config;

> server {
>         listen 443 ssl;
>         root /config/www/fqdn-rediretion;
>         index index.php;
>         server_name fqdn-rediretion;
> 
>         include /config/nginx/proxy.conf;
>         include /config/nginx/ssl_default.conf;
> 
>         location / {
>                 proxy_pass https://192.168.1.99:6443/;
>         }
> }
> 
> 
> 
> 
> 

With another working docker with my piwik instanace running in.

All three of my dockers are isolated and so I am unclear as your config appears to assume piwik and your website for redirection on contained in the same physical server or 1 docker?

root **{piwik_root_path};**
    index index.php;

    server_name **{shortener_url}**;

**Comment by mannp** [comment]
@mgazdzik me again...

I have tried going back to basics so I am not in conflict with the other webservers I have setup.

I have a nginx docker server with php configured and the following default config;

server {
listen 80 default_server;
listen 443 ssl;
root /config/www;
index index.php;
server_name _;
rewrite "^/([a-zA-Z0-9]{6})$" https://piwik-fqdn/piwik/index.php?module=APImethod=ShortcodeTracker.performRedirectForShortcode&code=$1;

    location / {
            index index.php;
    }

    ssl_certificate /config/keys/cert.crt;
    ssl_certificate_key /config/keys/cert.key;
    client_max_body_size 0;

    location ~ \.php$ {
            try_files $uri =404;
            fastcgi_split_path_info ^(.+\.php)(/.+)$;
            fastcgi_pass 127.0.0.1:9000;
            fastcgi_index index.php;
            fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
            include /etc/nginx/fastcgi_params;
    }

}

With an nginx server acting as a proxy with the following default config;

server {
listen 443 ssl;
root /config/www/fqdn-rediretion;
index index.php;
server_name fqdn-rediretion;

    include /config/nginx/proxy.conf;
    include /config/nginx/ssl_default.conf;

    location / {
            proxy_pass https://192.168.1.99:6443/;
    }

}

With another working docker with my piwik instanace running in.

All three of my dockers are isolated and so I am unclear as your config appears to assume piwik and your website for redirection on contained in the same physical server or 1 docker?

root {piwik_root_path};
index index.php;

server_name {shortener_url};

**Comment by mgazdzik** [comment]
Hi there @mannp,
From what I know it's possible to run proxying webserver (the one which is
hosting your shortener domain) and Piwik on another physical machine.
The only thing that you must "teach" to proxying webserver is what's
address of Piwik instance and redirect there the shortcode you caught in
nginx directive.
Actually by putting {shortener_url} I do mean to put full url where your
Piwik resides. that's not a variable :)

Also I can't see your messages for which I'm receiving notifications - not
sure if GH bug or you just deleting me. ;) please feel free to drop me a
line on my personal email or maybe we can jump on a quick call sometime to
sort your problems out?

Cheers :)

**Comment by mannp** [comment]
Hi @mgazdzik lol, not deleting you, but deleting my previous post to have one last try myself before I bother you again! :)

I have three working components but they do not work together.

(0) nginx proxy (with a vhost per docker below);
                   -> (1) nginx docker with piwik
                   -> (2) nginx docker for website
                   -> (3) nginx docker for nextcloud
                   -> etc
                   -> *(4) *nginx docker with your nginx config for redirects**

My confusion is that even if I redirect the shortcode thorugh (0) to (4) the nginx config of (4) has no connection to the piwik container at all, so piwik_root_path and shortener_url are not known within the redirects nginx docker;

root {piwik_root_path};
server_name {shortener_url};

If I was to add the redirect to my website container, it again has no knowledge of piwik_root_path and shortener_url;

So I was going to try to add fqdn.piwik/short to my piwik docker and see if I can get that to work, even though that would not be my ideal config.

**Comment by mannp** [comment]
@mgazdzik Thank you very much for your personal support over the weekend, allowing me to see the error in my vhost file.

Now everything works as it should and I can use your excellent plugin.

Thanks again for your support and for developing the plugin! :)

**Comment by mgazdzik** [comment]
Glad to hear, hope you'll enjoy plugin even more along the way :)
At any time in future please feel free to send feedback/support request, I'll be more than happy to support :)
Cheers!
